TROUBLESHOOTING.md

```diff
@@ -2,13 +2,116 @@
+
+### State management
+
+One of core features of Terraform is storing [infrastructure state][terraform-state]. The state can
+either be stored [locally, or remotely][state-storage]. The state contains mapping of existing
+infrastructure and it serves multiple purposes: drift detection, performance, additional metadata
+tracking.
+
+Local state is usually stored in `terraform.tfstate` file, while remote state storage specifics depend on
+[state backend][state-backend] being used. `terraform state` subcommand provides CLI capabilities
+for state management.
+
+*Note:* Terraform state may contain [sensitive data][state-sensitive-data].
+
+### Terraform native logging
+
+Terraform itself, independently of the provider plugin being used, has some debugging options that
+can enable additional logging. They are documented in [Terraform debugging
+page][terraform-debugging].
+
+### Provider version constraints
+
+Terraform providers are installed automatically by using `terraform init` command. The plugin is
+fetched based on [`required_providers` section within `terraform` configuration
+block][provider-versioning]. An important part of the configuration block is [version
+constraint][provider-version-constraints] that defines which version of the plugin should be
+installed.
+
+Terraform provider plugins follow [semantic versioning][semver] pattern and
+[MAJOR.MINOR.PATCH][provider-semver].
+
+```
+terraform {
+  required_providers {
+    vcfa = {
+      source = "vmware/vcfa"
+      version = "~> 1.0.0" # pins major and minor versions, but will accept new patch versions (e.g. 1.0.1)
+    }
+  }
+}
+
+provider "vcfa" {
+  # Configuration options
+}
+```
+
+### Terraform architecture and responsibility boundaries
+
+Terraform consists of three main components:
+* Core (`terraform` binary)
+* Plugins (called providers) e.g. Terraform Provider VCFA
+* Upstream APIs (some Go SDK for the platform)
+
+#### Terraform Core responsibilities
+
+The **Core** that is best identified by `terraform` binary in the consumer's system, is developed by
+Hashicorp. It provides:
+
+* HCL (Hashicorp Configuration Language) engine and syntax
+* Schema diff that is being shown when performing operations
+* Dependency graph for the entities
+* All the tooling, including:
+  * Terraform enterprise
+  * Workspaces
+  * CDK
+  * Registry and documentation format
+* Some fields in any resource:
+  * `lifecycle`
+  * `provisioner`
+
+#### Terraform Provider plugin responsibilities
+
+Plugin responsibilities, **Terraform provider VCFA** in this case:
+
+* Communication with the platform API (VCFA)
+* Implementation of provider plugin using Terraform plugin SDK
+  * Architecture (entity schema and granularity of resources/data sources)
+  * Implementation of Create, Read, Update, Delete and Import (CRUD+I for each resource
+  * Implementation of Read (R) for each data source
+* Maintenance of Go SDK
+* Releasing the provider to Terraform registry
```

Comment:
Let's add a link to registry (even though it's not there yet).

Reply:
Added